File: 0002-add-two-numbers/0002-add-two-numbers.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val = 0, _next = nil)
#     @val = val
#     @next = _next
#   end
# end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  prev = ListNode.new
  curr1, curr2, curr, carry = l1, l2, prev, 0
  while curr1 || curr2
    val1 = curr1 ? curr1.val : 0
    val2 = curr2 ? curr2.val : 0
    curr.next = ListNode.new(val1+val2+carry)
    if curr1
      curr1 = curr1.next
    end
    if curr2
      curr2 = curr2.next
    end
    curr = curr.next
    carry = curr.val / 10
    curr.val %= 10
  end
  if carry > 0
    curr.next = ListNode.new(carry)
  end
  return prev.next
end

File: 0002-add-two-numbers/0002-add-two-numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
  prev, curr1, curr2, carry := ListNode{ Val: 0, Next: nil, }, l1, l2, 0
  curr := &prev
  for curr1 != nil || curr2 != nil {
    val1, val2 := 0, 0
    if curr1 != nil {
      val1 = curr1.Val
    }
    if curr2 != nil {
      val2 = curr2.Val
    }
    next := ListNode{
      Val: val1 + val2 + carry,
      Next: nil,
    }
    curr.Next = &next
    if curr1 != nil {
      curr1 = curr1.Next
    }
    if curr2 != nil {
      curr2 = curr2.Next
    }
    curr = curr.Next
    carry = curr.Val / 10
    curr.Val %= 10
  }
  if carry > 0 {
    next := ListNode{
      Val: carry,
      Next: nil,
    }
    curr.Next = &next
  }
  return prev.Next
}